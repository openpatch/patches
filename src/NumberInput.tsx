import { ChangeEvent, KeyboardEvent } from "react";
import { css } from "@emotion/react";
import { Box } from "./Box";
import { StyledInput } from "./Input";
import { ChevronDown, ChevronUp } from "./icons/outline";

export type NumberInputProps = {
  id?: string;
  max?: number;
  min?: number;
  step?: number;
  value?: number;
  onChange?: (n: number | null) => void;
  onBlur?: () => void;
  disabled?: boolean;
  error?: boolean;
  required?: boolean;
};

export const NumberInput = ({
  id,
  min,
  max,
  step = 1,
  value = 0,
  onChange = () => {},
  onBlur = () => {},
  disabled = false,
  error = false,
  required = false,
}: NumberInputProps) => {
  function handleKeyDown(e: KeyboardEvent<HTMLInputElement>) {
    let flag = true;
    switch (e.key) {
      case "ArrowUp":
        increase();
        break;
      case "ArrowDown":
        decrease();
        break;
      case "Home":
        toMin();
        break;
      case "End":
        toMax();
        break;
      default:
        flag = false;
    }

    if (flag) {
      e.stopPropagation();
      e.preventDefault();
    }
  }

  function handleInnerChange(v: number | null) {
    if (!disabled) {
      onChange(v);
    }
  }

  function handleChange(e: ChangeEvent<HTMLInputElement>) {
    const n = Number(e.currentTarget.value);
    if (Number.isNaN(n)) {
      handleInnerChange(null);
    } else {
      handleInnerChange(n);
    }
  }

  function toMax() {
    if (max) {
      handleInnerChange(max);
    }
  }

  function toMin() {
    if (min) {
      handleInnerChange(min);
    }
  }

  function increase() {
    const newValue = value + step;
    if (!max || newValue <= max) {
      handleInnerChange(newValue);
    }
  }

  function decrease() {
    const newValue = value - step;
    if (!min || newValue >= min) {
      handleInnerChange(newValue);
    }
  }

  return (
    <Box position="relative">
      <StyledInput
        id={id}
        required={required}
        disabled={disabled}
        error={error}
        padding="xsmall"
        css={css`
          box-sizing: border-box;
          appearance: none;
          margin: 0;
          width: 100%;
          position: relative;
          padding-left: 1rem;
          padding-right: 1rem;
        `}
        onChange={handleChange}
        value={value?.toString()}
        onKeyDown={handleKeyDown}
        onBlur={onBlur}
        type="text"
        inputMode="decimal"
        pattern="[0-9]*(.[0-9]+)?"
        role="spinbutton"
      />
      <Box
        zIndex="10"
        cursor="pointer"
        position="absolute"
        borderLeftColor="neutral.100"
        borderLeftStyle="solid"
        css={css`
          top: 1px;
          bottom: 1px;
          right: 2px;
          border-left-width: 1px;
        `}
        width="24px"
      >
        <Box
          aria-label={`+${step}`}
          height="50%"
          width="24px"
          textAlign="center"
          top={0}
          backgroundColor="neutral.50"
          position="absolute"
          role="button"
          tabIndex={-1}
          css={css`
            border-bottom-width: 1px;
          `}
          borderBottomColor="neutral.100"
          borderBottomStyle="solid"
          onClick={increase}
        >
          <ChevronUp height="1em" />
        </Box>
        <Box
          aria-label={`-${step}`}
          height="50%"
          bottom={0}
          width="24px"
          textAlign="center"
          backgroundColor="neutral.50"
          position="absolute"
          role="button"
          tabIndex={-1}
          onClick={decrease}
        >
          <ChevronDown height="1em" />
        </Box>
      </Box>
    </Box>
  );
};
