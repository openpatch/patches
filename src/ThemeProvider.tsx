import React, { ReactElement } from "react";
import { ThemeProvider as EThemeProvider } from "emotion-theming";
import { Theme, base } from "./themes";
import { GlobalCssTheme } from "./themes/utils";

interface ThemeProviderProps {
  theme?: Theme;
  children: ReactElement;
}

const ThemeProvider = ({ theme = base, children }: ThemeProviderProps) => {
  return (
    <EThemeProvider theme={theme}>
      <GlobalCssTheme theme={theme} />
      {children}
    </EThemeProvider>
  );
};

export default ThemeProvider;
