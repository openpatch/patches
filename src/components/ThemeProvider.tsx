import React, { ReactElement } from "react";
import {
  ThemeProvider as SCThemeProvider,
  DefaultTheme,
} from "styled-components";
import defaultTheme from "../shared/theme";

interface ThemeProviderProps {
  theme?: DefaultTheme;
  children: ReactElement;
}

const ThemeProvider = ({
  theme = defaultTheme,
  children,
}: ThemeProviderProps) => {
  return <SCThemeProvider theme={theme}>{children}</SCThemeProvider>;
};

export default ThemeProvider;
