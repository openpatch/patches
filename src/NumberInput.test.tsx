
import { render } from "@testing-library/react";
import { NumberInput, NumberInputProps } from "./NumberInput";
import { ThemeProvider } from "./ThemeProvider";

const renderWithTheme = (props?: NumberInputProps) =>
render(
    <ThemeProvider>
        <NumberInput {...props} />
    </ThemeProvider>
);

test("should render", () => {
   renderWithTheme();
});

test("should match snapshot", () => {
    const { container } = renderWithTheme();
    expect(container).toMatchSnapshot();
});