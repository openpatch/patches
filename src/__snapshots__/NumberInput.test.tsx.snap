// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
.emotion-0 {
  box-sizing: border-box;
  min-width: 0;
  -webkit-transition: all 0.2s ease;
  transition: all 0.2s ease;
  position: relative;
}

.emotion-1 {
  border-radius: 4px;
  padding: 8px 4px;
  border-color: #E4E7EB;
  border-style: solid;
  border-width: 1px;
  box-sizing: border-box;
  width: 100%;
  padding: 8px;
  box-sizing: border-box;
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  margin: 0;
  width: 100%;
  position: relative;
  padding-left: 1rem;
  padding-right: 1rem;
}

.emotion-1::-webkit-input-placeholder {
  color: #7B8794;
}

.emotion-1::-moz-placeholder {
  color: #7B8794;
}

.emotion-1:-ms-input-placeholder {
  color: #7B8794;
}

.emotion-1::placeholder {
  color: #7B8794;
}

.emotion-1:focus {
  outline-color: transparent;
  outline-width: 0;
  border-color: #0C6B58;
  box-shadow: 0 0 0 3px rgba(66, 153, 225, 0.5);
  z-index: 1;
}

.emotion-1:disabled {
  background-color: #F5F7FA;
}

.emotion-2 {
  box-sizing: border-box;
  min-width: 0;
  -webkit-transition: all 0.2s ease;
  transition: all 0.2s ease;
  z-index: 10;
  cursor: pointer;
  position: absolute;
  border-left-color: #E4E7EB;
  border-left-style: solid;
  width: 24px;
  top: 1px;
  bottom: 1px;
  right: 2px;
  border-left-width: 1px;
}

.emotion-3 {
  box-sizing: border-box;
  min-width: 0;
  -webkit-transition: all 0.2s ease;
  transition: all 0.2s ease;
  height: 50%;
  width: 24px;
  text-align: center;
  top: 0;
  background-color: #F5F7FA;
  position: absolute;
  border-bottom-color: #E4E7EB;
  border-bottom-style: solid;
  border-bottom-width: 1px;
}

.emotion-4 {
  box-sizing: border-box;
  min-width: 0;
  -webkit-transition: all 0.2s ease;
  transition: all 0.2s ease;
  height: 50%;
  bottom: 0;
  width: 24px;
  text-align: center;
  background-color: #F5F7FA;
  position: absolute;
}

<div>
  <div
    id="patches-root"
  >
    <div
      class="emotion-0"
    >
      <input
        class="emotion-1"
        inputmode="decimal"
        pattern="[0-9]*(.[0-9]+)?"
        role="spinbutton"
        type="text"
        value="0"
      />
      <div
        class="emotion-2"
        cursor="pointer"
        width="24px"
      >
        <div
          aria-label="+1"
          class="emotion-3"
          height="50%"
          role="button"
          tabindex="-1"
          width="24px"
        >
          <svg
            fill="none"
            height="1em"
            stroke="currentColor"
            viewBox="0 0 24 24"
          >
            <path
              d="M5 15l7-7 7 7"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
          </svg>
        </div>
        <div
          aria-label="-1"
          class="emotion-4"
          height="50%"
          role="button"
          tabindex="-1"
          width="24px"
        >
          <svg
            fill="none"
            height="1em"
            stroke="currentColor"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 9l-7 7-7-7"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
</div>
`;
